app/src/main/java/com/sun/chat_04/data/remote/MessageRemoteDataSource.kt

```diff
@@ -1,39 +1,96 @@
 package com.sun.chat_04.data.remote
 
+import android.graphics.Bitmap
 import com.google.firebase.auth.FirebaseAuth
 import com.google.firebase.database.DataSnapshot
 import com.google.firebase.database.DatabaseError
 import com.google.firebase.database.FirebaseDatabase
 import com.google.firebase.database.ValueEventListener
+import com.google.firebase.storage.FirebaseStorage
 import com.sun.chat_04.data.model.Friend
 import com.sun.chat_04.data.model.Message
 import com.sun.chat_04.data.repositories.MessageDataSource
 import com.sun.chat_04.ui.signup.RemoteCallback
 import com.sun.chat_04.util.Constants
+import java.io.ByteArrayOutputStream
+import java.text.DateFormat
+import java.util.Date
 
 class MessageRemoteDataSource(
     auth: FirebaseAuth,
     private val database: FirebaseDatabase,
-    friend: Friend
+    friend: Friend,
+    private val storage: FirebaseStorage
 ) :
     MessageDataSource.Remote {
 
     private val uid = auth.currentUser?.uid.toString()
     private val uidUserRec = friend.idUser
-    override fun insertMessage(message: Message, callback: RemoteCallback<Boolean>) {
+    private val userSendRef = "${Constants.MESSAGES}/$uid/$uidUserRec"
+    private val userRecRef = "${Constants.MESSAGES}/$uidUserRec/$uid"
+    private lateinit var bytes: ByteArray
+    private val outputStream = ByteArrayOutputStream()
 
-        val userSendRef = "${Constants.MESSAGES}/$uid/$uidUserRec"
-        val userRecRef = "${Constants.MESSAGES}/$uidUserRec/$uid"
+    override fun insertMessage(message: Message, callback: RemoteCallback<Boolean>) {
         val idMessage = database.reference.child(userSendRef)
             .child(userRecRef).push().key
-        val bodyMessage = HashMap<String, String>()
-        bodyMessage[Constants.CONTENTS] = message.contents
-        bodyMessage[Constants.FROM] = uid
-        bodyMessage[Constants.ID_MESSAGE] = idMessage.toString()
-        val detailMessage = HashMap<String, Any>()
-        detailMessage["$userRecRef/$idMessage"] = bodyMessage
-        detailMessage["$userSendRef/$idMessage"] = bodyMessage
-        database.reference.updateChildren(detailMessage)
+        if (message.type == Constants.TEXT_MESSAGE) {
+            updateMessageDetail(
+                idMessage?.let {
+                    Message(it, message.contents, uid, Constants.TEXT_MESSAGE)
+                }
+
+            )
+        } else {
+            uploadImageToStorage(idMessage, callback)
+        }
+    }
+
+    private fun uploadImageToStorage(idMessage: String?, callback: RemoteCallback<Boolean>) {
```

Comment:
function upload image mà k có image thì sao upload đây nhỉ?
```suggestion
    private fun uploadImageToStorage(byte: ByteArrray, callback: RemoteCallback<Boolean>) {
```